onadata/libs/serializers/xform_serializer.py

```diff
@@ -142,12 +152,33 @@ def get_enketo_url(self, obj):
 
             url = self._get_metadata(obj, 'enketo_url')
             if url is None:
-                url = _create_enketo_url(self.context.get('request'), obj)
+                enketo_urls = _create_enketo_urls(
+                    self.context.get('request'), obj)
+                url = enketo_urls.get('offline_url')
 
             return _set_cache(ENKETO_URL_CACHE, url, obj)
 
         return None
 
+    def get_enketo_single_submit_url(self, obj):
+        if obj:
+            _enketo_single_submit_url = cache.get(
+                '{}{}'.format(ENKETO_SINGLE_SUBMIT_URL_CACHE,
+                              obj.pk))
+            if _enketo_single_submit_url:
+                return _enketo_single_submit_url
+
+            url = self._get_metadata(obj, 'enketo_url')
+            if url is None:
+                enketo_urls = _create_enketo_urls(
```

Comment:
I see that the `_create_enketo_urls` function is called on each of the `get_enketo_<url_type>` function... Does this generation of new urls invalidate the other urls ? _I'm wondering if calling it in multiple places might be a problem.... Is it possible for use to have one function set all the urls then this function just calls it if it's not set or retrieves it from metadata?_

Reply:
> I see that the _create_enketo_urls function is called on each of the get_enketo_<url_type> function


Yes, unfortunately, the XForm serializer sets the fields as MethodFields. To generate the value of these fields, different methods are invoked.


> Does this generation of new urls invalidate the other urls ?

I dont believe that the fetch can invalidate the other urls. The enketo urls are generated in Enketo and are not regenerated in Ona, merely fetched.

> I'm wondering if calling it in multiple places might be a problem

I dont think there is any underlying issue(functionality wise) with having this called in multiple places. The values are not overridden or such, but I agree, it does create duplicates in the code, which doesnt look all too awesome.


>  Is it possible for us to have one function set all the urls then this function just calls it if it's not set(re-setting all the possible enketo URLs) or retrieves it from metadata?

I believe that is what `create_enketo_urls` function is doing? Or be more detailed kiasi hapo

Reply:
Mhm okay... On the "one function" comment, I meant having the `MetaData` objects for all URLs set in one function and have the `get_enketo<url_type>` function do exactly what the name suggests; get the URL and call something else to reset if not present.

https://github.com/onaio/onadata/pull/1857#discussion_r475417301

Reply:
Which as you've mentioned `create_enketo_urls` already does... Which makes me wonder whether the extra calls to create the MetaData again like this one are needed ?